### PR TITLE
Map Mqtt5MessageException to Mqtt3MessageException for Mqtt3 client

### PR DESCRIPTION
**Pull request by david-katz**

### CODE REVIEW

**Comment by david-katz** [comment]
I have tried it for Connect. Is this a viable way of doing the mapping?

File: src/main/java/org/mqttbee/api/mqtt/mqtt3/exceptions/Mqtt3MessageException.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.mqttbee.api.mqtt.mqtt3.exceptions;
+
+import org.mqttbee.annotations.NotNull;
+
+/**
+ * @author David Katz
+ */
+public class Mqtt3MessageException extends Exception {
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/api/mqtt/mqtt3/exceptions/Mqtt3MessageException.java]: This class needs a method "Mqtt3Message getMqttMessage()" similar to the Mqtt5MessageException

File: src/main/java/org/mqttbee/mqtt/mqtt3/exceptions/Mqtt3ExceptionFactory.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+package org.mqttbee.mqtt.mqtt3.exceptions;
+
+import org.mqttbee.api.mqtt.mqtt3.exceptions.Mqtt3MessageException;
+import org.mqttbee.api.mqtt.mqtt5.exceptions.Mqtt5MessageException;
+
+public class Mqtt3ExceptionFactory {
+    public static Throwable map(Throwable throwable) {
+        if (Mqtt5MessageException.class.isAssignableFrom(throwable.getClass())) {
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/mqtt3/exceptions/Mqtt3ExceptionFactory.java]: "throwable instanceof Mqtt5MessageException" would be a lot easier

File: src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java [hunk 3]
```diff
@@ -54,15 +60,17 @@ public Mqtt3ClientView(@NotNull final Mqtt5ClientImpl wrapped) {
     @Override
     public Single<Mqtt3ConnAck> connect(@NotNull final Mqtt3Connect connect) {
         final Mqtt3ConnectView connectView =
-                MustNotBeImplementedUtil.checkNotImplemented(connect, Mqtt3ConnectView.class);
-        return wrapped.connect(connectView.getWrapped()).map(Mqtt3ConnAckView::create);
+            MustNotBeImplementedUtil.checkNotImplemented(connect, Mqtt3ConnectView.class);
+        return wrapped.connect(connectView.getWrapped())
+            .onErrorResumeNext(e -> e instanceof Mqtt5MessageException ? Single.error(new Mqtt3MessageException(e.getMessage())) : Single.error(e))
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java]: use your Mqtt3ExceptionFactory

Comment (david-katz) [reply src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java]: done

[the remark this answers is not in the document]
Comment (david-katz) [reply src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java]: done

File: src/main/java/org/mqttbee/mqtt/mqtt3/exceptions/Mqtt3ExceptionFactory.java (hunk 2 above)
Comment (david-katz) [reply src/main/java/org/mqttbee/mqtt/mqtt3/exceptions/Mqtt3ExceptionFactory.java]: done

File: src/main/java/org/mqttbee/api/mqtt/mqtt3/exceptions/Mqtt3MessageException.java (hunk 1 above)
Comment (david-katz) [reply src/main/java/org/mqttbee/api/mqtt/mqtt3/exceptions/Mqtt3MessageException.java]: done, but not currently able to call constructor with Mqtt3Message

File: src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java [hunk 4]
```diff
@@ -29,23 +28,57 @@
...
  * @author Silvio Giebl
+ * @author David Katz
  */
 public class Mqtt3ClientView implements Mqtt3Client {
 
     private final Mqtt5ClientImpl wrapped;
 
+    private static final Function<? super Throwable, ? extends Single<? extends Mqtt5ConnAck>>
+        mapMqtt5toMqtt3ExceptionForConnect =
+        e -> e instanceof Mqtt5MessageException ? Single.error(Mqtt3ExceptionFactory.map(e)) : Single.error(e);
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java]: e -> Single.error(Mqtt3ExceptionFactory.map(e)) is sufficient as Mqtt3ExceptionFactory performs the instanceof check

File: src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java [hunk 5]
```diff
@@ -54,49 +87,63 @@ public Mqtt3ClientView(@NotNull final Mqtt5ClientImpl wrapped) {
     @Override
     public Single<Mqtt3ConnAck> connect(@NotNull final Mqtt3Connect connect) {
         final Mqtt3ConnectView connectView =
-                MustNotBeImplementedUtil.checkNotImplemented(connect, Mqtt3ConnectView.class);
-        return wrapped.connect(connectView.getWrapped()).map(Mqtt3ConnAckView::create);
+            MustNotBeImplementedUtil.checkNotImplemented(connect, Mqtt3ConnectView.class);
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java]: Why is the indentation wrong?

File: src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java [hunk 6]
```diff
@@ -29,23 +28,57 @@
...
 
 /**
  * @author Silvio Giebl
+ * @author David Katz
  */
 public class Mqtt3ClientView implements Mqtt3Client {
 
     private final Mqtt5ClientImpl wrapped;
 
+    private static final Function<? super Throwable, ? extends Single<? extends Mqtt5ConnAck>>
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java]: Function<Throwable, Single<Mqtt5ConnAck>> is more readable

File: src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java [hunk 7]
```diff
@@ -29,23 +28,57 @@
...
+    private static final Function<? super Throwable, ? extends Publisher<? extends Mqtt5Publish>>
+        mapMqtt5toMqtt3ExceptionsForRemainingPublishes =
+        e -> e instanceof Mqtt5MessageException ? Flowable.error(Mqtt3ExceptionFactory.map(e)) : Flowable.error(e);
+
+    private static final Function<? super Throwable, ? extends CompletableSource>
+        mapMqtt5toMqtt3ExceptionsForDisconnect =
+        e -> e instanceof Mqtt5MessageException ? Completable.error(Mqtt3ExceptionFactory.map(e)) :
+            Completable.error(e);
+
+    private static final Function<? super Throwable, ? extends Publisher<? extends Mqtt5Publish>>
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java]: same as mapMqtt5toMqtt3ExceptionsForRemainingPublishes

File: src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java [hunk 8]
```diff
@@ -29,23 +28,57 @@
...
 /**
  * @author Silvio Giebl
+ * @author David Katz
  */
 public class Mqtt3ClientView implements Mqtt3Client {
 
     private final Mqtt5ClientImpl wrapped;
 
+    private static final Function<? super Throwable, ? extends Single<? extends Mqtt5ConnAck>>
+        mapMqtt5toMqtt3ExceptionForConnect =
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/mqtt3/Mqtt3ClientView.java]: naming convention for static fields: uppercase
=> MQTT5_TO_3_EXCEPTION_CONNACK_MAPPER

**Pull request merged**